library/alloc/src/vec/mod.rs

```diff
@@ -1969,6 +2001,49 @@ pub fn from_elem_in<T: Clone, A: Allocator>(elem: T, n: usize, alloc: A) -> Vec<
     <T as SpecFromElem>::from_elem(elem, n, alloc)
 }
 
+trait ExtendFromWithinSpec {
+    /// Safety: src need to be valid index
+    unsafe fn spec_extend_from_within(&mut self, src: Range<usize>);
+}
+
+impl<T: Clone, A: Allocator> ExtendFromWithinSpec for Vec<T, A> {
+    default unsafe fn spec_extend_from_within(&mut self, src: Range<usize>) {
+        self.reserve(src.len());
+        let i = {
+            // copied from `spare_capacity_mut` b/c it doesn't allow access to the content
+            let spare_space = unsafe {
+                slice::from_raw_parts_mut(
+                    self.as_mut_ptr().add(self.len) as *mut MaybeUninit<T>,
+                    self.buf.capacity() - self.len,
+                )
+            };
+
+            // Safety:
+            // - caller guaratees that src is a valid index
+            let to_clone = unsafe { self.get_unchecked(src) };
```

Comment:
I'm a bit unsure whether this is valid or not directly on `self`, given our mutable borrow of the spare capacity above. It might be better to slice this off the borrow returned by the new `split_at_spare_mut` method, which will have the same bounds as `self` did.